clusterloader2/pkg/measurement/common/nodelocaldns_latency_prometheus.go

```diff
@@ -0,0 +1,123 @@
+)
+
+func init() {
+	create := func() measurement.Measurement {
+		return CreatePrometheusMeasurement(&nodelocaldnsLatencyGatherer{})
+	}
+	if err := measurement.Register(nodelocaldnsLatencyPrometheusMeasurementName, create); err != nil {
+		klog.Fatalf("Cannot register %s: %v", nodelocaldnsLatencyPrometheusMeasurementName, err)
+	}
+}
+
+type nodelocaldnsLatencyGatherer struct{}
+
+func (n *nodelocaldnsLatencyGatherer) Gather(executor QueryExecutor, startTime time.Time, config *measurement.Config) ([]measurement.Summary, error) {
+	result, err := n.getPercentileLatencies(executor, startTime)
+	if err != nil {
+		return nil, err
+	}
+	content, err := util.PrettyPrintJSON(result)
+	if err != nil {
+		return nil, err
+	}
+	summaries := []measurement.Summary{measurement.CreateSummary(n.String(), "json", content)}
+	return summaries, n.validateResult(config, result)
+}
+
+func (n *nodelocaldnsLatencyGatherer) String() string {
+	return nodelocaldnsLatencyPrometheusMeasurementName
+}
+
+func (n *nodelocaldnsLatencyGatherer) IsEnabled(config *measurement.Config) bool {
+	return true
+}
+
+func (n *nodelocaldnsLatencyGatherer) validateResult(config *measurement.Config, result *latencyPrometheus) error {
+	latencyUpperBound, err := util.GetFloat64OrDefault(config.Params, "dns_latency_upper_bound", float64(defaultLatencyUpperBoundSeconds))
+	if err != nil {
+		return err
+	}
+	if result.Perc99 > latencyUpperBound {
+		return errors.NewMetricViolationError(
+			"NodelocalDNS dns_request_duration_seconds",
+			fmt.Sprintf("99thPercentile Latency %f is higher than the upper bound of %f", result.Perc99, latencyUpperBound))
+	}
+	return nil
+}
+
+func (n *nodelocaldnsLatencyGatherer) getPercentileLatencies(executor QueryExecutor, startTime time.Time) (*latencyPrometheus, error) {
+	endTime := time.Now()
+	measurementDuration := endTime.Sub(startTime)
+	promDuration := measurementutil.ToPrometheusTime(measurementDuration)
+	var errStr []string
+	results := make(map[string]float64)
+	for _, percVal := range desiredPercentiles {
+		percDecimal := fmt.Sprintf("0.%d", percVal)
+		query := fmt.Sprintf(percLatencyQueryTemplate, percDecimal, promDuration)
+		samples, err := executor.Query(query, endTime)
+		if err != nil {
+			errStr = append(errStr, fmt.Sprintf("failed to execute query %q, err - %v", query, err))
+			continue
+		}
+		if len(samples) != 1 {
+			errStr = append(errStr, fmt.Sprintf("got unexpected number of samples: %d for query %q", len(samples), query))
+			continue
+		}
+		results[fmt.Sprintf("perc%d", percVal)] = float64(samples[0].Value)
+	}
+	if len(errStr) > 0 {
+		return nil, fmt.Errorf("failed to compute latencies, errors - %v", errStr)
+	}
+	jsonResult, err := json.Marshal(results)
+	if err != nil {
+		return nil, fmt.Errorf("failed to marshal result %v into json, err - %v", results, err)
+	}
+	l := &latencyPrometheus{}
+	err = json.Unmarshal(jsonResult, l)
+	return l, err
+}
+
+type latencyPrometheus struct {
```

Comment:
Let's use https://github.com/kubernetes/perf-tests/blob/39b9f8f5569282a5ad858c7a852c2b2324fd6d13/clusterloader2/pkg/measurement/util/latency_metric.go#L29
instead

[and then you don't need marshaling/unmarshaling - just use SetQuantile method]

Reply:
didn't know this existed, thanks for pointing out. Done.